server/src/main/java/org/elasticsearch/http/HttpClientStatsTracker.java

```diff
@@ -0,0 +1,185 @@
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Adjust the stats for the given channel to reflect the latest request received.
+     */
+    void updateClientStats(final HttpRequest httpRequest, final HttpChannel httpChannel) {
+        if (clientStatsEnabled && httpChannel != null) {
+            HttpStats.ClientStats clientStats = httpChannelStats.get(getChannelKey(httpChannel));
+            if (clientStats == null) {
+                // will always return a non-null value when httpChannel is non-null
+                clientStats = addClientStats(httpChannel);
+            }
+
+            if (clientStats.agent == null) {
+                final String elasticProductOrigin = getFirstValueForHeader(httpRequest, "x-elastic-product-origin");
+                if (elasticProductOrigin != null) {
+                    clientStats.agent = elasticProductOrigin;
+                } else {
+                    final String userAgent = getFirstValueForHeader(httpRequest, "User-Agent");
+                    if (userAgent != null) {
+                        clientStats.agent = userAgent;
+                    }
+                }
+            }
+            if (clientStats.localAddress == null) {
+                clientStats.localAddress =
+                    httpChannel.getLocalAddress() == null ? null : NetworkAddress.format(httpChannel.getLocalAddress());
+                clientStats.remoteAddress =
+                    httpChannel.getRemoteAddress() == null ? null : NetworkAddress.format(httpChannel.getRemoteAddress());
+            }
+            if (clientStats.forwardedFor == null) {
+                final String forwardedFor = getFirstValueForHeader(httpRequest, "x-forwarded-for");
+                if (forwardedFor != null) {
+                    clientStats.forwardedFor = forwardedFor;
+                }
+            }
+            if (clientStats.opaqueId == null) {
+                final String opaqueId = getFirstValueForHeader(httpRequest, "x-opaque-id");
+                if (opaqueId != null) {
+                    clientStats.opaqueId = opaqueId;
+                }
+            }
+            clientStats.lastRequestTimeMillis = threadPool.absoluteTimeInMillis();
+            clientStats.lastUri = httpRequest.uri();
+            clientStats.requestCount.increment();
+            clientStats.requestSizeBytes.add(httpRequest.content().length());
+        }
+    }
+
+    /**
+     * @return a list of the stats for the channels that are currently being tracked.
+     */
+    List<HttpStats.ClientStats> getClientStats() {
+        pruneClientStats(false);
+        return new ArrayList<>(httpChannelStats.values());
+    }
+
+    /**
+     * Returns a key suitable for use in a hash table for the specified HttpChannel
+     */
+    private static int getChannelKey(HttpChannel channel) {
+        // always use an identity-based hash code rather than one based on object state
+        return System.identityHashCode(channel);
```

Comment:
Unrelated comment but isn't this broken? I guess it's approximately ok and that might be good enough for the stats, but shouldn't we have a collision safer key here for the stats map?